### ISSUE TITLE
Questions about the Centris algorithm implementation

### ISSUE THREAD

**Issue by Tricker-z** [WOOSEUNGHOON/Centris-public#1]
Hi, Centris devs:

I encountered some questions while trying to understand the Centris source code logic. Here are part of the code snippets in `Preprocessor_full.py` (line 292-301).

```python
for S_hashval in verDateDict[S]:
	score = tlsh.diffxlen(hashval, S_hashval)
	if int(score) <= 30:
		if verDateDict[S][hashval] == "NODATE" or verDateDict[OSS][hashval] == "NODATE":
			candiX[OSS] += 1
			temp[OSS].append(hashval)
			
		elif verDateDict[OSS][hashval] <= verDateDict[S][hashval]:
			candiX[OSS] += 1
			temp[OSS].append(hashval)
```

Based on my knowledge of the algorithm in the paper, I think `S` here should use `S_hashval` instead of `hashval` to compare the birth time with `OSS`, i.e., `verDateDict[OSS][hashval] <= verDateDict[S][S_hashval]`.

Additionally, line 312 in `Preprocessor_full.py`:

```python
(float(candiX[X])/float(aveFuncs[X])) >= theta
```

I would like to know why `aveFuncs` is used here instead of `allFuncs`.

 I would greatly appreciate any help or hint that can be provided :D

**Comment by WOOSEUNGHOON** [comment]
Hi,

1. You're right. `hashval` should be changed to `S_hashval`. There was a mistake in the process of refactoring the code. I will 
check and correct it. Thank you for your report.

2. `allFuncs` contains all functions in every version of X. However, if S reuses X, only functions corresponding to one (or a few) version will be reused; thus `aveFuncs` is used to determine the similarity based on the average number of functions belonging to each version of X.

**Comment by Tricker-z** [comment]
Hi, Centris devs:

   Thanks for your reply, while for the second issue, i.e., `allFuncs` or `aveFuncs` should be utilized, I still think it should be `allFuncs`.  According to the following code snippets in `Preprocessor_full.py` (line 277).

```python
for eachVal in jsonStr:
  hashval = eachVal['hash']
  ...

for X in candiX:
  ...
  if (float(candiX[X])/float(aveFuncs[X])) >= theta:
    ...
```

Here we traverse all the functions (in all the versions and marked with hash value) contained in an OSS. Assume that we have a great number of versions while each verison contains just a few functions. In this case, if we use `aveFuncs`, the ratio `(float(candiX[X])/float(aveFuncs[X]))` can be very large :(

In the first loop, we consider all the functions in all the versions, there is no reason that we use the average functions to calculate the ratio. Any further discussion is very welcome.

**Comment by WOOSEUNGHOON** [comment]
Hi, 

Of course, the situation you mentioned may happen.

However, we thought that if an OSS project is reused, the number of reused functions is generally fewer than the number of functions belonging to one version (even if we traverse all functions in every version of an OSS). Therefore, we determined that it would be more accurate to use the average number of functions in each X version rather than measuring the ratio based on the total number of unique functions in all versions of X.

Also, the opposite case of what you say may exist. For example, when X contains an enormous number of versions, the ratio can be very small if we use `allFuncs`, even though a sufficient amount of functions of a version of X are being reused in S.
Although we did not measure the differences between using `aveFuncs` and `allFuncs` by experiment, when we examined the cases at the time of the paper experiment, many of the aforementioned cases appeared, thus we implemented Centris using `aveFuncs`. This is part of algorithm implementation; if we use `allFuncs`, more OSS will be identified as the prime OSS (i.e., do not reuse any other OSS), and thus false negatives will be reduced in detection results (but this may increase false positives).  

In summary, we agree with your concern, but we used `aveFuncs` because we found that it is more efficient in component detection.

**Comment by Tricker-z** [comment]
Hi,

Thanks very much. I think I get the idea of your design.

I have an additional question to ask, did you find any problems with the incorrect birth time of the functions in OSS during your experiments. For instance, if the project was migrated, it may cause the birth time to be wrong.  I think this may be a common situation, especially in library projects.  This problem should be serious because it can cause errors in the entire code segmentation.

**Comment by wangjg33** [comment]
> Hi,
> 
> Thanks very much. I think I get the idea of your design.
> 
> I have an additional question to ask, did you find any problems with the incorrect birth time of the functions in OSS during your experiments. For instance, if the project was migrated, it may cause the birth time to be wrong. I think this may be a common situation, especially in library projects. This problem should be serious because it can cause errors in the entire code segmentation.

-----------------------------------------------------------------------------------------------------------------------------------------
hi!
  Nice to meet you. I am also reproducing this project. Coincidentally, I am also in Shenzhen. Can you leave your email address and let us talk about relevant issues.

**Comment by WOOSEUNGHOON** [comment]
@Tricker-z 

Hi.
That's a very good question.
As you said, issues that came from the wrong birth time are sometimes founded in our dataset (not frequently);
owing to such issues, some false results occurred in code segmentation.
In the experiment, we ignored such issues and performed code segmentation, and therefore some false alarms occurred 
(false positives/negatives caused by incorrect code segmentation mentioned in the paper).

The most accurate way would be to manually identify OSS versions' birth dates whenever such false alarms occur, 
which is too heuristic and requires a lot of effort. 

Although we ignored such errors as the effect was insignificant, 
you may get more accurate results if you use a more reliable birth time instead of using the release date of the GitHub tag.




@wangjg33 

Hi,
I answered your questions by email.
Please check it.

Thank you.

**Comment by Tricker-z** [comment]
Hi,

Nice to meet you, we can further discuss the project here

wangjg33 ***@***.***> 于2022年7月13日周三 17:51写道：

**Comment by Tricker-z** [comment]
> > Hi,
> > Thanks very much. I think I get the idea of your design.
> > I have an additional question to ask, did you find any problems with the incorrect birth time of the functions in OSS during your experiments. For instance, if the project was migrated, it may cause the birth time to be wrong. I think this may be a common situation, especially in library projects. This problem should be serious because it can cause errors in the entire code segmentation.
> 
> hi! Nice to meet you. I am also reproducing this project. Coincidentally, I am also in Shenzhen. Can you leave your email address and let us talk about relevant issues.

Hi

Nice to meet you. You can check my profile to get the email address. We can discuss further related issues.

**Comment by Tricker-z** [comment]
> @Tricker-z
> 
> Hi. That's a very good question. As you said, issues that came from the wrong birth time are sometimes founded in our dataset (not frequently); owing to such issues, some false results occurred in code segmentation. In the experiment, we ignored such issues and performed code segmentation, and therefore some false alarms occurred (false positives/negatives caused by incorrect code segmentation mentioned in the paper).
> 
> The most accurate way would be to manually identify OSS versions' birth dates whenever such false alarms occur, which is too heuristic and requires a lot of effort.
> 
> Although we ignored such errors as the effect was insignificant, you may get more accurate results if you use a more reliable birth time instead of using the release date of the GitHub tag.
> 
> @wangjg33
> 
> Hi, I answered your questions by email. Please check it.
> 
> Thank you.

Hi

**Comment by Tricker-z** [comment]
@WOOSEUNGHOON 

Hi

Thanks very much for your reply. I have no further questions about the project and I will close the issue.

Best regards
